bpo-43976: add vendor config

FFY00 opened the pull request:
Signed-off-by: Filipe Laíns <lains@riseup.net>

Python-Dev proposal:
https://gist.github.com/FFY00/625f65681fbcd7fc039dd4d727bb2c2f
https://mail.python.org/archives/list/python-dev@python.org/thread/OWETD7ZHCHX3R4IMGO3YPRC6WABVWGXH/

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-43976](https://bugs.python.org/issue43976) -->
https://bugs.python.org/issue43976
<!-- /issue-number -->

FFY00 committed:
bpo-41282: add vendor config

Signed-off-by: Filipe Laíns <lains@riseup.net>

pradyunsg commented on Doc/library/site.rst:
```diff
@@ -34,7 +34,7 @@ are skipped.  For the tail part, it uses the empty string and then
 :file:`lib/python{X.Y}/site-packages` (on Unix and Macintosh), and finally
 the ``purelib`` and ``platlib`` paths for each scheme specified in the
-``EXTRA_SITE_INSTALL_SCHEMES`` variable of the vendor config.  For each
+``EXTRA_SITE_INSTALL_SCHEMES`` listr variable of the vendor config. For each
```
There's a typo!

FFY00 replied:
Thanks :sweat_smile:

jaraco commented on Lib/test/test_site.py:
```diff
@@ -277,6 +284,11 @@ def test_getusersitepackages(self):
+        # force re-load of vendor schemes with the patched sys.path
+        site._VENDOR_SCHEMES = None
+        sysconfig._load_vendor_schemes()
+
```
Would it make sense to have a separate test rather than mutate an existing test?

FFY00 replied:
Sure, we still need the test as-is, but we can also keep the original one.

jaraco commented on Lib/test/test_sysconfig.py:
```diff
@@ -263,10 +278,14 @@ def test_get_config_h_filename(self):
+        # force re-load of vendor schemes with the patched sys.path
+        sysconfig._load_vendor_schemes()
+
+        wanted = ['nt', 'posix_home', 'posix_prefix', 'some_vendor']
```
Would it make sense to have two tests, one which patches sys.path and another that doesn't?

FFY00 replied:
Ditto.

jaraco reviewed (changes requested):
This looks good to me and a good start for providing hooks for vendor configurability.

I'd like to see an answer to @merwok's questions:

> Can you build python from scratch if sysconfig imports site?
> What about cross-compilation? (which has unclear support status)

I'm not familiar with building from scratch, but I agree, it would be important to address these cases. Another way to think about it - is this change backward-compatible? Can this change provide a graceful degradation such that if no vendor file is supplied, the behavior is all-but-guaranteed to be the same as on prior versions (if no `--vendor-config` is supplied)? If the change is "safe" in that respect (compatibility), the feature can be refined through the betas.

@merwok Do you know how one could test for building Python from scratch (to prove compatibility or demonstrate a failure)?

Also, the docs builds are failing - so those checks will need to be fixed before merging.

Those things addressed, I'm ready to say this change is suitable for merge, acknowledging that the short time frame prior to the feature freeze means this change may have a fair likelihood of being reverted (with all the consternation that brings and leaving no opportunity to restore prior to Python 3.11).

hroncok commented on Lib/sysconfig.py:
```diff
@@ -58,6 +58,25 @@
+        _INSTALL_SCHEMES.update({
+            name: scheme
+            for name, scheme in _vendor_config.EXTRA_INSTALL_SCHEMES.items()
+            if name not in _INSTALL_SCHEMES
```
If this check was not here, we could use this instead of our patch entirely. I'm not sure what is the benefit of the check.

FFY00 replied:
That breaks several assumptions in other parts of the code, so I chose to not go into that right now. We could remove it in the future if it there is a valid use case, though I can't really think of anything strong enough to warrant that.

jaraco replied:
FFY00, are you saying that removing the check would break several assumptions in other parts of the code? Can you provide an example of such a broken assumption? Isn't "vendor doesn't have to patch sysconfig" a compelling reason to warrant it (given that's the goal of this patch)?

FFY00 replied:
> FFY00, are you saying that removing the check would break several assumptions in other parts of the code? Can you provide an example of such a broken assumption?

There are parts of the `site`, `venv` and `distutils` modules, for example, that hardcode `site-packages`. Allowing vendors to change this path would cause at the minimum a moderate amount of breakage.

> Isn't "vendor doesn't have to patch sysconfig" a compelling reason to warrant it (given that's the goal of this patch)?

It is, though, generally there are several parts of the code that can be change to solve such issues, and I think there might be easier, and probably better, ways to do it rather than allowing vendors to change the default install schemes.

So the question here should be, why would vendors need to change the default schemes? Can we solve the issue in a different way? What would make more sense, removing this check, or any of the alternatives?

Fedora's case can be solved by changing `distutils.command.install` to respect `get_preferred_schemes`, which I feel is a much better solution.
If there are any other reasons where someone might want to remove this check, we can look at them and decide what makes the most sense.

vstinner commented on Lib/site.py:
```diff
@@ -74,6 +74,7 @@
 import builtins
 import _sitebuiltins
 import io
+import sysconfig
```
sysconfig creates tons of Python objects and is slow to import. I would prefer to avoid it at startup.

Would it be possible to move the logic into a smaller module which does exactly what you need, and reuse this module in sysconfig? See _sitebuiltins for example.

FFY00 replied:
Yes, I think this should be perfectly doable. It also stacks up another reason to implement @zooba's suggestion of making sysconfig a directory and installing the vendor config as a submodule, to prevent shadowing it.

uranusjr replied:
> It also stacks up another reason to implement @zooba's suggestion of making sysconfig a directory and installing the vendor config as a submodule, to prevent shadowing it.

Maybe I’m misunderstanding, but are you hinting to put the smaller module in a subpackage? Since that won’t work because `import sysconfig.bare` implicitly imports `sysconfig` as well so unless you move all names out of `sysconfig` (not an option since it breaks all existing code) this will not help import time.

FFY00 replied:
Right :facepalm:

vstinner replied:
you can add a `_my_bootstrap_module` module and import it in `sysconfig` and `site` modules. It doesn't change the API, only the implementation, and Python startup time: https://pythondev.readthedocs.io/startup_time.html

tiran commented on Doc/Makefile:
```diff
@@ -6,8 +6,8 @@
-SPHINXBUILD  = PATH=$(VENVDIR)/bin:$$PATH sphinx-build
-BLURB        = PATH=$(VENVDIR)/bin:$$PATH blurb
+SPHINXBUILD  = $(VENVDIR)/bin/python $(VENVDIR)/bin/sphinx-build
+BLURB        = $(VENVDIR)/bin/python $(VENVDIR)/bin/blurb
```
This looks like an unrelated change and breaks building of documentation w/o virtual envs

FFY00 replied:
It is, see [`aa36fb5` (#25718)](https://github.com/python/cpython/pull/25718/commits/aa36fb5850e366ab7d0225faaa09f8080cb6d251), I don't think it's required anymore.

tiran commented on Doc/library/site.rst:
```diff
@@ -32,14 +32,20 @@ It starts by constructing up to four directories from a head and a tail part.
 .. versionchanged:: 3.5
    Support for the "site-python" directory has been removed.
 
+.. versionchanged:: 3.10
```
It's a new feature, so 3.11.
```suggestion
.. versionchanged:: 3.11
```

tiran commented on aclocal.m4:
```diff
@@ -1,4 +1,4 @@
-# generated automatically by aclocal 1.16.2 -*- Autoconf -*-
+# generated automatically by aclocal 1.16.3 -*- Autoconf -*-
```
Please use https://quay.io/repository/tiran/cpython_autoconf to regenerate the autoconf files.

FFY00 committed:
update generated files

Signed-off-by: Filipe Laíns <lains@riseup.net>

FFY00 committed:
implement lazy loading of attributes in _sysconfig

Signed-off-by: Filipe Laíns <lains@riseup.net>

FFY00 committed:
note _sysconfig/sysconfig split in the module docstrings

Signed-off-by: Filipe Laíns <lains@riseup.net>

FFY00 committed:
fix _PROJECT_BASE

Signed-off-by: Filipe Laíns <lains@riseup.net>

FFY00 committed:
forgot to add this to the main commit

I fixed the main commit in another branch, but forgot I am no longer
rebasing here .-.

Signed-off-by: Filipe Laíns <lains@riseup.net>

FFY00 committed:
fix _PROJECT_BASE again

Signed-off-by: Filipe Laíns <lains@riseup.net>

FFY00 committed:
fix _SRCDIR

Signed-off-by: Filipe Laíns <lains@riseup.net>

FFY00 committed:
freeze _sysconfig

Signed-off-by: Filipe Laíns <lains@riseup.net>
